### DISCUSSION TITLE
Return annotation of ``response.json()``

### DISCUSSION THREAD

**Discussion by jonathanslenders** [encode/httpx#1666] [Ideas]
Right now, the function signature of ``response.json()`` returns ``Any``.

The issue with this is that mypy won't complain when we write:
```python
data = await response.json()
```
The result is not awaitable. This is common however, when migrating from other HTTP clients like aiohttp.

Ideally, I'd like the return type to be something like this:
```python
-> Union[str, int, dict, list]
```

I'm not sure whether that's actually feasible.

**Comment by lovelydinosaur** [comment]
We'd want `-> Union[str, int, float, bool, dict, list]`, right?

Switching to that would probably be okay. It's not complete (because we're not fully expressing the types of dict and list\*), but it is sufficient to catch this kind of error.

*\* Note that Python **can't** fully express the type that `.json()` returns.*
